core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -47,9 +49,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val inputMetrics = blockResult.inputMetrics
         val existingMetrics = context.taskMetrics
           .getInputMetricsForReadMethod(inputMetrics.readMethod)
-        existingMetrics.addBytesRead(inputMetrics.bytesRead)
+        existingMetrics.incBytesRead(inputMetrics.bytesRead)
 
-        new InterruptibleIterator(context, blockResult.data.asInstanceOf[Iterator[T]])
+        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        new InterruptibleIterator(context, AfterNextInterceptingIterator(iter, (next: T) => {
+          existingMetrics.incRecordsRead(1)
```

Comment:
@ksakellis I think Patrick means locking in the sense that `AtomicLong` and its ilk use special compare-and-swap hardware instructions that lock the bus.  They're more performant than acquiring a software lock, but still have high overhead compared to updating a variable that's in CPU cache.

http://en.wikipedia.org/wiki/Compare-and-swap